packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -299,6 +300,19 @@ class _DefaultTabControllerState extends State<DefaultTabController> with Single
     );
   }
 
+  @override
+  void didUpdateWidget(covariant DefaultTabController oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.length != oldWidget.length) {
+      final int oldIndex = _controller._index;
+      _controller = TabController(
```

Comment:
Need to dispose the old controller first, but: what about clients that still hold a reference to the old controller?

Reply:
Opened https://github.com/flutter/flutter/pull/24783